java/memory/src/main/java/io/netty/buffer/ArrowBuf.java

```diff
@@ -17,144 +17,84 @@
 
 package io.netty.buffer;
 
-import java.io.IOException;
-import java.io.InputStream;
-import java.io.OutputStream;
 import java.nio.ByteBuffer;
 import java.nio.ByteOrder;
-import java.nio.channels.FileChannel;
-import java.nio.channels.GatheringByteChannel;
-import java.nio.channels.ScatteringByteChannel;
-import java.nio.charset.Charset;
 import java.util.concurrent.atomic.AtomicInteger;
 import java.util.concurrent.atomic.AtomicLong;
 
-import org.apache.arrow.memory.AllocationManager.BufferLedger;
-import org.apache.arrow.memory.ArrowByteBufAllocator;
-import org.apache.arrow.memory.BaseAllocator;
+import io.netty.util.IllegalReferenceCountException;
+import io.netty.util.internal.MathUtil;
+import org.apache.arrow.memory.*;
 import org.apache.arrow.memory.BaseAllocator.Verbosity;
-import org.apache.arrow.memory.BoundsChecking;
-import org.apache.arrow.memory.BufferAllocator;
-import org.apache.arrow.memory.BufferManager;
 import org.apache.arrow.memory.util.HistoricalLog;
 import org.apache.arrow.util.Preconditions;
 
 import io.netty.util.internal.PlatformDependent;
 
 /**
- * ArrowBuf is the abstraction around raw byte arrays that
- * comprise arrow data structures.
+ * ArrowBuf serves as a facade over underlying memory by providing
+ * several access APIs to read/write data into a chunk of direct
+ * memory. All the accounting, ownership and reference management
+ * is done by {@link ReferenceManager} and ArrowBuf can work
+ * with a custom user provided implementation of ReferenceManager
  *
+ * Two important instance variables of an ArrowBuf:
  *
- * <p>Specifically, it serves as a facade over
- * {@linkplain UnsafeDirectLittleEndian} memory objects that hides the details
- * of raw memory addresses.
+ * (1) address - starting virtual address in the underlying memory
+ * chunk that this ArrowBuf has access to
  *
- * <p>ArrowBuf supports reference counting and ledgering to closely track where
- * memory is being used.
+ * (2) length - length (in bytes) in the underlying memory chunk
```

Comment:
should this be tied to memory?  What if we want a file backed implementation?

Reply:
It definitely should be tied to a memory address. Whether that is backed by a memory mapped file will be up to the consumer with this change.